File: providers/aws/src/main/java/cloudspec/aws/ec2/loader/EC2CapacityReservationLoader.java
/*-
 * #%L
 * CloudSpec AWS Provider
 * %%
 * Copyright (C) 2020 Ezequiel Foncubierta
 * %%
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 * #L%
 */
package cloudspec.aws.ec2.loader;

import cloudspec.aws.IAWSClientsProvider;
import cloudspec.aws.ec2.resource.EC2CapacityReservationResource;

import java.util.List;
import java.util.stream.Stream;

public class EC2CapacityReservationLoader extends EC2ResourceLoader<EC2CapacityReservationResource> {
    public EC2CapacityReservationLoader(IAWSClientsProvider clientsProvider) {
        super(clientsProvider);
    }

    @Override
    protected Stream<EC2CapacityReservationResource> getResourcesInRegion(String region,
                                                                          List<String> ids) {
        try (var client = clientsProvider.getEc2ClientForRegion(region)) {
            // https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribeCapacityReservations.html
            var response = client.describeCapacityReservations(builder ->
                    builder.capacityReservationIds(
                            ids.toArray(new String[0])
                    )
            );

            return response.capacityReservations()
                           .stream()
                           .map(capacityReservation -> EC2CapacityReservationResource.fromSdk(region, capacityReservation));
        }
    }
}

File: providers/aws/src/main/kotlin/cloudspec/aws/ec2/EC2CapacityReservationLoader.kt
/*-
 * #%L
 * CloudSpec AWS Provider
 * %%
 * Copyright (C) 2020 Ezequiel Foncubierta
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
package cloudspec.aws.ec2

import cloudspec.aws.IAWSClientsProvider
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.withContext
import kotlin.streams.toList

class EC2CapacityReservationLoader(clientsProvider: IAWSClientsProvider) :
        EC2ResourceLoader<EC2CapacityReservation>(clientsProvider) {

    override suspend fun resourcesInRegion(region: String,
                                           ids: List<String>): List<EC2CapacityReservation> = coroutineScope {
        clientsProvider.ec2ClientForRegion(region).use { client ->
            withContext(Dispatchers.Default) {
                client
                    // https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribeCapacityReservations.html
                    .describeCapacityReservations { builder ->
                        if (ids.isNotEmpty()) {
                            builder.capacityReservationIds(ids)
                        }
                    }
                    .capacityReservations()
                    .stream()
                    .map { it.toEC2CapacityReservation(region) }
                    .toList()
            }
        }
    }
}
